### PR TITLE
Publish platform binaries to Azure storage

### PR DESCRIPTION
**Pull request by kichalla**

### CODE REVIEW

File: build/buildBuildpacksImages.sh [hunk 1]
```diff
@@ -4,18 +4,12 @@
...
+set -ex
 
-# Enables running from within other scripts that already declared $REPO_DIR
-if [ -z "$REPO_DIR" ]; then
+if [ -z $REPO_DIR ]; then
 	declare -r REPO_DIR=$( cd $( dirname "$0" ) && cd .. && pwd )
 fi
 
-# Enables running from within other scripts that already sourced __variables.sh
-if [ -z "$__REPO_DIR" ]; then
```
Comment (kichalla) [inline build/buildBuildpacksImages.sh]: Will revert this particular change.

File: build/buildPlatformBinaries.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
```
Comment (kichalla) [inline build/buildPlatformBinaries.sh]: FYI:

image


image

image

File: images/build/Dockerfile [hunk 3]
```diff
@@ -106,34 +105,34 @@ RUN apt-get update \
...
- && ~/n/bin/n -d 12.9.1 \
- && ~/n/bin/n -d $NODE6_VERSION \
- && mv /usr/local/n/versions/node /opt/nodejs \
- && rm -rf /usr/local/n ~/n
+COPY images/build/__common.sh /tmp/scripts
+COPY images/build/installPlatform.sh /tmp/scripts
+RUN chmod +x /tmp/scripts/installPlatform.sh
+RUN cd /tmp/scripts \ 
+ && . ./__nodeVersions.sh \
+ && ./installPlatform.sh nodejs 4.4.7 \
```
Comment (kichalla) [inline images/build/Dockerfile]: So we are installing the binaries that we put in storage. As of this PR I am doing this only in build image. Will send out a separate PR for updating the runtime images.

File: images/build/__common.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    local downloadedFileName="$3"
+    local headersFile="/tmp/headers.txt"
+
+    echo "Downloading $platformName version '$version'..."
+    curl \
+        -D $headersFile \
+        -SL "https://oryxsdksdev.blob.core.windows.net/$platformName/$platformName-$version.tar.gz" \
+        --output $downloadedFileName
+
+    headerName="x-ms-meta-checksum"
```
Comment (kichalla) [inline images/build/__common.sh]: checksum verification

File: images/build/Dockerfile [hunk 5]
```diff
@@ -177,22 +176,14 @@ RUN set -ex \
...
 FROM main AS python
 # https://github.com/docker-library/python/issues/147
 ENV PYTHONIOENCODING UTF-8
 COPY build/__pythonVersions.sh /tmp/scripts
-COPY --from=py27-build-base /opt /opt
-COPY --from=py36-build-base /opt /opt
+RUN cd /tmp/scripts \
+    && . ./__pythonVersions.sh \
+    && ./installPlatform.sh python $PYTHON27_VERSION \
+    && ./installPlatform.sh python $PYTHON36_VERSION
```
Comment (arroyc) [inline images/build/Dockerfile]: python 3.8 and 3.7?

File: images/build/__common.sh [hunk 6]
```diff
@@ -0,0 +1,19 @@
+function downloadFileAndVerifyChecksum() {
+    local platformName="$1"
+    local version="$2"
+    local downloadedFileName="$3"
+    local headersFile="/tmp/headers.txt"
+
+    echo "Downloading $platformName version '$version'..."
+    curl \
+        -D $headersFile \
+        -SL "https://oryxsdksdev.blob.core.windows.net/$platformName/$platformName-$version.tar.gz" \
```
Comment (arroyc) [inline images/build/__common.sh]: nit: keep this link 'https://oryxsdksdev.blob.core.windows.net'  in constant.yaml?

File: platforms/__common.sh [hunk 7]
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (arroyc) [inline platforms/__common.sh]: nit: can we add some suffix to the file name as I see two common.sh? it may be easier to maintain

File: platforms/node/versionsToBuild.txt [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+10.14.2
+10.16.3
+10.18.0
+12.9.1
+12.11.0
+12.11.1
+12.12.0
+12.13.0
+12.14.0
+
```
Comment (arroyc) [inline platforms/node/versionsToBuild.txt]: nit: blank line?

File: platforms/php/buildPhp.sh [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+		echo
+
+		# Installing PHP composer requires having PHP installed in an first image first, so we try installing
+		# a version here.
+		docker build \
+			-f "$phpPlatformDir/composer/Dockerfile" \
+			--build-arg PHP_VERSION="7.3.5" \
+			--build-arg PHP_SHA256="e1011838a46fd4a195c8453b333916622d7ff5bce4aca2d9d99afac142db2472" \
+			--build-arg GPG_KEYS="CBAF69F173A0FEA4B537F470D66C9593118BCCB6 F38252826ACD957EF380D39F2F7956BC5DA04B5D" \
+			--build-arg COMPOSER_VERSION="$version" \
```
Comment (arroyc) [inline platforms/php/buildPhp.sh]: can we also keep these version, sha and key in constants.yaml?

File: images/build/Dockerfile (hunk 5 above)
Comment (kichalla) [reply images/build/Dockerfile]: That is being installed in the slim image. Check `slim.Dockerfile`

File: images/build/__common.sh (hunk 6 above)
Comment (kichalla) [reply images/build/__common.sh]: Good point
